File: go/series/series.go
package series

// All returns a list of all substrings of s with length n.
func All(n int, s string) []string {
	length := 1 + (len(s) - n)
	result := make([]string, length)
	for i := 0; i < length; i++ {
		result[i] = s[i : i+n]
	}
	return result
}

// UnsafeFirst returns the first substring of s with length n.
func UnsafeFirst(n int, s string) string {
	return s[:n]
}

// First safely returns the first substring of s with length n.
func First(n int, s string) (first string, ok bool) {
	if n > len(s) {
		return "", false
	}
	return s[:n], true
}

File: c/series/src/series.c
#include "series.h"
#include <stdlib.h>
#include <string.h>

slices_t slices(char *input_text, unsigned int substring_length)
{
    slices_t result = {0};

    int len = strlen(input_text);
    if (len == 0 || substring_length == 0)
    {
        return result;
    }
    result.substring_count = 1 + (len - substring_length);
    result.substring = calloc(sizeof(char *), result.substring_count);

    for (int i = 0; i <= (int)(len - substring_length); i++)
    {
        result.substring[i] = calloc(sizeof(char), substring_length);
        strncpy(result.substring[i], input_text + i, substring_length);
    }
    return result;
}
